src/StoreApi/Utilities/QuantityLimits.php

```diff
@@ -0,0 +1,211 @@
+	 * @param string $rounding_function ceil, floor, or round.
+	 * @return int
+	 */
+	public function limit_to_multiple( int $number, int $multiple_of, string $rounding_function = 'round' ) {
+		if ( $multiple_of <= 1 ) {
+			return $number;
+		}
+		$rounding_function = in_array( $rounding_function, [ 'ceil', 'floor', 'round' ], true ) ? $rounding_function : 'round';
+		return $rounding_function( $number / $multiple_of ) * $multiple_of;
+	}
+
+	/**
+	 * Check that a given quantity is valid according to any limits in place.
+	 *
+	 * @param integer           $quantity Quantity to validate.
+	 * @param \WC_Product|array $cart_item Cart item.
+	 * @return \WP_Error|true
+	 */
+	public function validate_cart_item_quantity( $quantity, $cart_item ) {
+		$limits = $this->get_cart_item_quantity_limits( $cart_item );
+
+		if ( ! $limits['editable'] ) {
+			return new \WP_Error(
+				'readonly_quantity',
+				__( 'This item is already in the cart and it\'s quantity cannot be edited', 'woo-gutenberg-products-block' )
+			);
+		}
+
+		if ( $quantity < $limits['minimum'] ) {
+			return new \WP_Error(
+				'invalid_quantity',
+				sprintf(
+					// Translators: %s amount.
+					__( 'The minimum quantity that can be added to the cart is %s', 'woo-gutenberg-products-block' ),
+					$limits['minimum']
+				)
+			);
+		}
+
+		if ( $quantity > $limits['maximum'] ) {
+			return new \WP_Error(
+				'invalid_quantity',
+				sprintf(
+					// Translators: %s amount.
+					__( 'The maximum quantity that can be added to the cart is %s', 'woo-gutenberg-products-block' ),
+					$limits['maximum']
+				)
+			);
+		}
+
+		if ( $quantity % $limits['multiple_of'] ) {
+			return new \WP_Error(
+				'invalid_quantity',
+				sprintf(
+					// Translators: %s amount.
+					__( 'The quantity added to the cart must be a multiple of %s', 'woo-gutenberg-products-block' ),
+					$limits['multiple_of']
+				)
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Get the limit for the total number of a product allowed in the cart.
+	 *
+	 * This is based on product properties, including remaining stock, and defaults to a maximum of 99 of any product
+	 * in the cart at once.
+	 *
+	 * @param \WC_Product $product Product instance.
+	 * @return int
+	 */
+	protected function get_product_quantity_limit( \WC_Product $product ) {
+		$limits = [ 99 ];
+
+		if ( $product->is_sold_individually() ) {
+			$limits[] = 1;
+		} elseif ( ! $product->backorders_allowed() ) {
+			$limits[] = $this->get_remaining_stock( $product );
+		}
+
+		/**
+		 * Filters the quantity limit for a product being added to the cart via the Store API.
+		 *
+		 * Filters the variation option name for custom option slugs.
+		 *
+		 * @param integer $quantity_limit Quantity limit which defaults to 99 unless sold individually.
+		 * @param \WC_Product $product Product instance.
+		 * @return integer
+		 */
+		return apply_filters( 'woocommerce_store_api_product_quantity_limit', max( min( array_filter( $limits ) ), 1 ), $product );
```

Comment:
I don't get `min( array_filter( $limits ) )`. Are you casting the filter value?

Reply:
I didn't change this https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/7329935b1dce1c8bab77b213ffb2742e088d3b13/src/StoreApi/Schemas/ProductSchema.php#L539 It's getting the min value from all possible limits, and then using max() to ensure it's greater than 1.